csrc/runtime/executor_kernel_arg.h

```diff
@@ -22,20 +22,48 @@
 
 namespace nvfuser {
 
+namespace detail {
+template <typename T>
+struct is_kernel_argument_holder : std::false_type {};
+
+template <>
+struct is_kernel_argument_holder<KernelArgumentHolder> : std::true_type {};
+
+template <typename T>
+inline constexpr bool is_kernel_argument_holder_v = is_kernel_argument_holder<
+    std::remove_cv_t<std::remove_reference_t<T>>>::value;
+
+template <typename... Args>
+inline constexpr bool no_kernel_argument_holder =
+    !(is_kernel_argument_holder_v<Args> || ...);
+} // namespace detail
+
 //! KernelArgumentHolder copies meta information from kernel inputs, including
 //! tensor sizes/shapes/dtype/memory_ptr and copies scalar inputs. It is used
-//! for both compilation as well as kernel execution. The important thing is to
-//! strip ownership of tensor from KernelArgumentHolder, so that during async
-//! compilation, we are not unnecessarily holding memory that is not needed.
-class KernelArgumentHolder {
+//! for both compilation as well as kernel execution. It takes ownership of
+//! at::Tensors so care should be taken when using it relative to tensor.
+class NVF_API KernelArgumentHolder {
  public:
   KernelArgumentHolder() = default;
 
   KernelArgumentHolder(const KernelArgumentHolder& self) = default;
 
-  NVF_API KernelArgumentHolder(
-      const c10::ArrayRef<c10::IValue>& inputs,
-      std::optional<int8_t> device = std::nullopt);
+  // Constructor using std::enable_if_t for C++17 compatibility to prevent
+  // implicit conversion to KernelArgumentHolder which can cause a recursive
```

Comment:
This can be avoided by:
```C++
KernelArgumentHolder(KernelArgumentHolder& self) = default;
```
Also, better to have the following as well:
```C++
KernelArgumentHolder(KernelArgumentHolder&& self) = default;
```

Reply:
Done.